source/extensions/filters/http/dynamic_modules/abi_impl.cc

```diff
@@ -197,6 +197,135 @@ bool envoy_dynamic_module_callback_http_get_response_trailers(
+  auto filter = static_cast<DynamicModuleHttpFilter*>(filter_envoy_ptr);
+  auto stream_info = filter->streamInfo();
+  if (!stream_info) {
+    ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::dynamic_modules), debug,
+                        "stream info is not available");
+    return nullptr;
+  }
+  auto metadata = stream_info->dynamicMetadata().mutable_filter_metadata();
+  absl::string_view namespace_view(static_cast<const char*>(namespace_ptr), namespace_length);
+  auto metadata_namespace = metadata->find(namespace_view);
+  if (metadata_namespace == metadata->end()) {
+    if (!create_if_not_exist) {
+      ENVOY_LOG_TO_LOGGER(
+          Envoy::Logger::Registry::getLog(Envoy::Logger::Id::dynamic_modules), debug,
+          fmt::format("namespace {} not found in dynamic metadata", namespace_view));
+      return nullptr;
+    }
+    metadata_namespace = metadata->emplace(namespace_view, ProtobufWkt::Struct{}).first;
+  }
+  return &metadata_namespace->second;
+}
+
+/**
+ * Helper to get the metadata value from the metadata namespace.
+ *
+ * @param filter_envoy_ptr is the pointer to the DynamicModuleHttpFilter object of the
+ * corresponding HTTP filter.
+ * @param namespace_ptr is the namespace of the dynamic metadata.
+ * @param namespace_length is the length of the namespace.
+ * @param key_ptr is the key of the dynamic metadata.
+ * @param key_length is the length of the key.
+ * @return the metadata value if it exists, nullptr otherwise.
+ */
+const ProtobufWkt::Value*
+getDynamicMetadataValue(envoy_dynamic_module_type_http_filter_envoy_ptr filter_envoy_ptr,
+                        envoy_dynamic_module_type_buffer_module_ptr namespace_ptr,
+                        size_t namespace_length,
+                        envoy_dynamic_module_type_buffer_module_ptr key_ptr, size_t key_length) {
+  auto metadata_namespace =
+      getDynamicMetadataNamespace(filter_envoy_ptr, namespace_ptr, namespace_length, false);
+  if (!metadata_namespace) {
+    return nullptr;
+  }
+  absl::string_view key_view(static_cast<const char*>(key_ptr), key_length);
+  auto key_metadata = metadata_namespace->fields().find(key_view);
+  if (key_metadata == metadata_namespace->fields().end()) {
+    ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::dynamic_modules), debug,
+                        fmt::format("key {} not found in metadata namespace", key_view));
+    return nullptr;
+  }
+  return &key_metadata->second;
+}
+
+bool envoy_dynamic_module_callback_http_set_dynamic_metadata_number(
+    envoy_dynamic_module_type_http_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_buffer_module_ptr namespace_ptr, size_t namespace_length,
+    envoy_dynamic_module_type_buffer_module_ptr key_ptr, size_t key_length, double value) {
+  auto metadata_namespace =
+      getDynamicMetadataNamespace(filter_envoy_ptr, namespace_ptr, namespace_length, true);
```

Comment:
How is the namespace created in the first place?

Reply:
clarified in https://github.com/envoyproxy/envoy/pull/37860/commits/8990fa25703586d6e9cd774fb0731f28ecda0d81 - when stream_info is available, setter functions will create it if not exist